### PR TITLE
Add support for djangorestframework-dataclasses

### PR DESCRIPTION
**Pull request by oxan**
Add an extension to support my [djangorestframework-dataclasses](https://github.com/oxan/djangorestframework-dataclasses) project. I've been getting a few requests for this: [#52](https://github.com/oxan/djangorestframework-dataclasses/issues/52), [#41](https://github.com/oxan/djangorestframework-dataclasses/issues/41), [#28](https://github.com/oxan/djangorestframework-dataclasses/issues/28).

### CODE REVIEW

**Comment by tfranzel** [comment]
Hi @oxan, 

thank you for the PR. this looks awesome. For live extensions we require at least one minimal test. See [contrib tests](https://github.com/tfranzel/drf-spectacular/tree/master/tests/contrib) for examples. Nothing fancy, just to check that basic functionality works as expected

Am I understanding correctly that your lib pretty much works out of the box except for the name and the default doc? Does it work similar to `ModelSerializer` where fields are build on demand, just with dataclasses instead of Models? Pretty neat!

**Comment by oxan** [comment]
> thank you for the PR. this looks awesome. For live extensions we require at least one minimal test. See [contrib tests](https://github.com/tfranzel/drf-spectacular/tree/master/tests/contrib?rgh-link-date=2022-01-23T21%3A06%3A35Z) for examples. Nothing fancy, just to check that basic functionality works as expected

I've added tests to the PR :)

> Am I understanding correctly that your lib pretty much works out of the box except for the name and the default doc? 

Yes.

> Does it work similar to `ModelSerializer` where fields are build on demand, just with dataclasses instead of Models? Pretty neat!

Yes, that's correct. I based my implementation upon `ModelSerializer`, actually (though there's been quite a bit of divergence since then).

File: tests/contrib/test_rest_framework_dataclasses.py [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+    name: str
+    leader: Person
+    members: typing.List[Person]
+
+
+class GroupSerializer(DataclassSerializer):
+    class Meta:
+        dataclass = Group
+
+
```
Comment (tfranzel) [inline tests/contrib/test_rest_framework_dataclasses.py:30]: to make the suite run through you probably need this also. the tests may skip but not break.

`@pytest.mark.skipif(sys.version_info < (3, 7), reason='dataclass required by package')`

File: tests/contrib/test_rest_framework_dataclasses.py [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+from dataclasses import dataclass
+from django.urls import path
+from drf_spectacular.utils import extend_schema
+from rest_framework.decorators import api_view
+from rest_framework_dataclasses.serializers import DataclassSerializer
+
+from tests import assert_schema, generate_schema
+
+
+@dataclass
```
Comment (tfranzel) [inline tests/contrib/test_rest_framework_dataclasses.py:13]: probably a good idea to move the dataclasses and the dataserializer inside the test (including the imports). we need to support 3.6 till April just as Django will.

File: drf_spectacular/contrib/rest_framework_dataclasses.py [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+    target_class = "rest_framework_dataclasses.serializers.DataclassSerializer"
+    match_subclasses = True
+
+    def get_name(self):
+        """Use the dataclass name in the schema, instead of the serializer prefix (which can be just Dataclass)."""
+        return self.target.dataclass_definition.dataclass_type.__name__
+
+    def strip_library_doc(self, schema):
+        """Strip the DataclassSerializer library documentation from the schema."""
+        if not hasattr(self, '_library_doc'):
```
Comment (tfranzel) [inline drf_spectacular/contrib/rest_framework_dataclasses.py:16]: good thinking, but there is no point in caching this in the instance, as a new instance is created on every new match.

**Comment by oxan** [comment]
The tests fail because `djangorestframework-dataclasses` isn't available on Python 3.6. How should I solve this? I can mark the dependency as 3.7+, but then the test will fail, as it tries to import `dataclasses`, which isn't available on 3.6.

**Review by tfranzel — COMMENTED** [review]
good stuff. just 2 minors to make the testsuite pass.

File: tests/contrib/test_rest_framework_dataclasses.py (hunk 1 above)
Comment (oxan) [reply tests/contrib/test_rest_framework_dataclasses.py:32]: Should all be fixed now.

File: drf_spectacular/contrib/rest_framework_dataclasses.py (hunk 3 above)
Comment (oxan) [reply drf_spectacular/contrib/rest_framework_dataclasses.py]: Ah, well, let's just extract it everytime then.

**Comment by tfranzel** [comment]
hehe, overlooked that one mistake too. makes sense of course.

**Comment by tfranzel** [comment]
sry, isort is picky like that... could you run ``python runtests.py` and then squash all commits. thanks!

**Comment by oxan** [comment]
Heh, I actually tried that, but it started complaining about a zillion library files, so I decided to roll the dice and hope CI didn't bite. Didn't realize until now that I could avoid that by moving the venv out of the my git clone.

Should be all good now!

**Comment by tfranzel** [comment]
Excellent! Thank you @oxan! I know the test suite can be a pain but that's how we sustain the quality.

p.s.: if the virtualenv is called `venv` it should have worked even if it is contained inside. the configs should ignore that case, which is how my setup works.

**Comment by oxan** [comment]
> I know the test suite can be a pain but that's how we sustain the quality.

No worries, I like well-tested software! It's just a bit of a pain that CI didn't run automatically on each push, otherwise I could've fixed it all myself in a few tries (or maybe I shouldn't have been too lazy to get the testsuite working locally :))

> p.s.: if the virtualenv is called venv it should have worked even if it is contained inside. the configs should ignore that case, which is how my setup works.

Ah okay, that make sense, I usually call my venvs `.venv` so it didn't ignore that.

**Comment by dineshtrivedi** [comment]
Hi @oxan @tfranzel , Thank you very much for this work!
Is there anything I can help to assist in releasing a new version including this fix?

**Comment by tfranzel** [comment]
@dineshtrivedi new version was just released. enjoy!

**Comment by dineshtrivedi** [comment]
Thank you! @tfranzel

**Pull request merged**